### PR TITLE
feat: new download csv popup with limit

### PR DESCRIPTION
**Pull request by rephus**
Closes:  #3571

### Description:

[Screencast from 25-10-22 16:51:16.webm](https://user-images.githubusercontent.com/1983672/197810591-63b30633-57e5-4190-9fe4-23f9eb76c985.webm)

### CODE REVIEW

File: packages/frontend/src/components/UnderlyingData/UnderlyingDataModalContent.tsx [hunk 1]
```diff
@@ -21,7 +21,10 @@ import { useParams } from 'react-router-dom';
 import { v4 as uuidv4 } from 'uuid';
 import { useExplore } from '../../hooks/useExplore';
 import { getExplorerUrlFromCreateSavedChartVersion } from '../../hooks/useExplorerRoute';
-import { useUnderlyingDataResults } from '../../hooks/useQueryResults';
+import {
+    getAllQueryResults,
```
Comment (IrakliJani) [inline packages/frontend/src/components/UnderlyingData/UnderlyingDataModalContent.tsx:25]: don't think this is used here anymore

File: packages/backend/src/routers/projectRouter.ts [hunk 2]
```diff
@@ -182,6 +182,39 @@ projectRouter.post(
...
+    isAuthenticated,
+    async (req, res, next) => {
+        try {
+            const { body } = req;
+            const metricQuery: MetricQuery = {
+                dimensions: body.dimensions,
+                metrics: body.metrics,
+                filters: body.filters,
+                sorts: body.sorts,
+                limit: 1000000,
```
Comment (IrakliJani) [inline packages/backend/src/routers/projectRouter.ts:208]: so there's still a limit, should we let users know the amount?

Comment (rephus) [reply packages/backend/src/routers/projectRouter.ts:208]: Yeah, this is something I'm not sure about... changing the limit to `undefined` will imply a few extra changes in the methods we use to run queries. In most DB the limit is measures in MB apparently. I found this on Stackoverflow: `MySQL suggests using 18446744073709551615 as the number of records in the limit` it is probably a maxInt or something ... but I don't think that limit is compatible with all databases... so I don't know, maybe 1000M ?

File: packages/frontend/src/components/DownloadCsvPopup/index.tsx [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        const csvRef = useRef<
+            CSVLink & HTMLAnchorElement & { link: HTMLAnchorElement }
+        >(null);
+        useEffect(() => {
+            if (csvRows.length > 0 && csvRef.current?.link) {
+                csvRef.current.link.click();
+                setCsvRows([]);
+                setIsOpen(false);
+            }
+        }, [csvRows]);
```
[the remark this answers is not in the document]
Comment (rephus) [reply packages/frontend/src/components/DownloadCsvPopup/index.tsx:55]: We normally do that when downloading the CSV, but in this case I want to trigger the download once my csvRows is set with the new data , and the only way I can achieve this is by using useEffect. I've tried with `done() async method' suggested by the library but I don't think it works with `useState` hooks. https://github.com/react-csv/react-csv/issues/189#issuecomment-818899971

**Comment by TuringLovesDeathMetal** [comment]
@rephus this looks awesome! 😍 

This wasn't in the spec, but looking at it, how complicated would it be to allow a `custom` option where you enter the number of rows you want to download? 

^If it's not easy, I'll open another issue for it!

File: packages/backend/src/routers/projectRouter.ts (hunk 2 above)
Comment (rephus) [reply packages/backend/src/routers/projectRouter.ts:208]: I'm going to do a refactor to specify an undefined limit

**Comment by rephus** [comment]
[Screencast from 26-10-22 11:47:49.webm](https://user-images.githubusercontent.com/1983672/197995585-b36d4c4d-3a49-4b0c-9241-aafb8e1966aa.webm)

**Review by IrakliJani — APPROVED** [review]
looks good!

**Comment by TuringLovesDeathMetal** [comment]
beautiful!

**Pull request merged**